### DISCUSSION TITLE
How can Cooja be run without the UI?

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2276] [Q&A] [answered]
Is there a way to run Cooja from the command line? If yes, what parameters can be specified for the run?

**Comment by pjonsson** [comment, marked as answer]
The option is called `-nogui=file.csc` and has been around for a while.

I don't recall when the build system was changed from ant to gradle, but I think Contiki-NG v4.8 Cooja understands `--help`. With gradle, here's how to show the help in the docker image: `cooja --args="--help"`. The option `-external_tools_config` does not do what you expect it to do, so that option is likely to change in the future.

Besides the seed, there isn't much that can be changed that isn't in the simulation file, but on latest master, you have per-simulation options (only one option implemented, https://github.com/contiki-ng/cooja/pull/780, and that is mainly relevant for `-quickstart`).

**Reply by rahav** [reply]
@pjonsson Thanks for your answer again. I'm on a pressed timeline with a paper for a conference, therefore I'm still running Contiki-NG v4.7, which uses `ant`. I hope you have a quick tip to make this work, because it doesn't for me.

From my project directory I ran `~/cooja -nogui=sims/my_sim_file.csc`, which compiles cooja and starts the GUI (without all of the regular windows but the blue background screen).

1. Assuming that this is a simple mistake and you can help me run it without the GUI. How do I specify the log filename I need the output to go to?
2. How do I specify how long the simulation should run?

Thank you,
Rahav

**Reply by pjonsson** [reply]
The version of Cooja you're using predates my involvement, -nogui has worked without creating any windows during the time I've been working on Cooja. It could be that `-nogui` requires the full path, so try:
```
-nogui=`pwd`/sims/my_sim_file.csc
```
I've heard that:
```
TIMEOUT(10000, log.testOK());
```
in the script is a common trick for getting a simulation to run a certain amount of time and then quit, but without failing. (You can see a variant of that in `tests/07-simulation-base/js/04-ringbufindex.js`, but the `log.testFailed()` shouldn't be required since a timeout is normally a test failure).

**Reply by rahav** [reply]
Thanks @pjonsson. In v4.8, can you specify on the command line the Simulation to run? Output log filename? and run duration?

Are there docs on how to automate Cooja runs?

**Reply by pjonsson** [reply]
Timelines are bit blurry to me, right now I have 1600 commits in Cooja and 375 commits in Contiki-NG since March, and the 4.8 release was in early July.

The command line handling was reworked for v4.8 to use Picocli according to the release notes (https://github.com/contiki-ng/contiki-ng/releases/tag/release%2Fv4.8). Effectively, this means Cooja is more tolerant for different argument orders, and became much better at telling you when there is something wrong with the arguments. I think the argument checking was further improved after v4.8, but it's more polish, not a huge improvement.

I don't recall when the AWT-shuffling was done, but that is just an improvement, it shouldn't be visible to you except that you get the backtrace dialog with a trace somewhere in the Java font library less often.

As for your first question, the log ends up in COOJA.log and COOJA.testlog. I think those files appear in the build directory with ant, but I'm not 100% certain, they could also end up in your current directory. Cooja in v4.9 will no longer generate COOJA.log by default (that is just what you see in the terminal), you will have to specify `--logname` to get that file generated.

The `-nogui` parameter is old, I'm not sure why it's not working as intended for you. Check `tests/07-simulation-base/*.csc` in Contiki-NG for examples of -nogui. (In the docker container, you should be able to type `make` in that directory to run the simulation tests, but I think the v4.7 Cooja is chatty and the build system redirects the output so it's hard to understand what is happening.)

**Reply by rahav** [reply]
Thank you!!!

and awesome number of commits... Thank you for improving this wonderful OS.

**Reply by arurke** [reply]
I'll just give a +1 to what @pjonsson is mentioning about the duration-trick, and getting inspiration from `tests/07-simulation-base/*.csc` (and its friends in v4.7: `tests/simexec.sh` and `tests/Makefile.simulation-test`) - that was my starting point when I set up a couple of years ago, and it definitely worked/works without UI.

**Reply by pjonsson** [reply]
This is very basic, but does `~/cooja` actually pass its parameters to Cooja? I have a recollection that I added support for that to some incarnation of the file past v4.7.

**Reply by arurke** [reply]
That's a good point, it does not look like it in my container ([~2 yr old](https://hub.docker.com/layers/contiker/contiki-ng/66cbc3434/images/sha256-deae47d07406e49a066da8f183722ace94762578a9e50b52387c801f77c881db?context=explore)):
```
user@a7b5bde01e13:~$ cat ~/cooja
#!/bin/bash
ant -Dbasedir=/home/user/contiki-ng/tools/cooja -f /home/user/contiki-ng/tools/cooja/build.xml run
```

**Reply by rahav** [reply]
I would like to propose that in v4.8 or soon beyond it, there will be rudimentary Cooja arguments that can be specified externally. It doesn't have to be command line; a configuration file will do, but basic operating parameters should be supported. We should probably open this for discussion and my list of basic is:

- sim to run
- seed (often it is very important to run with the same seed to repeat an experiment)
- with GUI
- run duration
- output log filename

Nice to have
- amount of memory

**Reply by pjonsson** [reply]
The Cooja script in v4.8 just passes on `"$@"', so it supports all of that except run duration.

**Reply by rahav** [reply]
Thanks @pjonsson, that is very encouraging. What is the best practice to end a non-interactive (no GUI) simulation in v4.8?

**Reply by pjonsson** [reply]
The tests in Contiki-NG are updated when Cooja changes, so they can be used as examples. The current API is `log.testOK()` and `log.testFailed()`, and I don't think that will be changed anytime soon.
